src/idl_gen_cpp.cpp

```diff
@@ -2717,10 +2725,22 @@ class CppGenerator : public BaseGenerator {
     for (auto it = struct_def.fields.vec.begin();
          it != struct_def.fields.vec.end(); ++it) {
       const auto &field = **it;
-      code_.SetValue("FIELD_TYPE",
-                     GenTypeGet(field.value.type, " ", "", " ", false));
+      const auto &field_type = field.value.type;
+
+      // Check if enum arrays are used without specifying --scoped-enums
+      if (!parser_.opts.scoped_enums && IsArray(field_type) &&
+          IsEnum(field_type.VectorType())) {
+        printf("--scoped-enums must be enabled to use enum arrays\n");
```

Comment:
Not happy with this error for multiple reasons. First it would be better if `--scoped-enums` wasn't required. And if that really isn't possible, I'd prefer to trigger this error in the parser, where are all error handling happens. The generators don't really have a way to do errors, and this print + return false is a bit of a hackey way to do it.

Reply:
@aardappel 
The idea to ignore untyped enums was mine. Initially, `Array` template had the support of enums without underlying type specification. But this third template parameter complicates the template and complicate compile-time helpers for the `Array` (compile-time functions like `size()` or other `constexpr` features).
There are a few C++ compilers without underlying type specification. I assumed that `Array` is a new feature needn't support of legacy. One should use a modern compiler to use `Array`.
Of course, this should be documented.

I can **restore** removed code if I'm not right.

Generator vs parser:
The flag `--scoped-enums` needed for C++ code generator only.
Other languages are independent of this flag.
One can use `flatc` to convert from json to binary and versa. 
The `--scoped-enums` isn't necessary for these conversions.

Another way is to silently add an underlying type to generated enums used with Array.
For example, the underlying type silently added for each int64/uint64 enum due to MSVC restrictions.

Reply:
I thought enums were always explicitly typed. But if they are not, they should IMO default to 32 bit like `int`. I actually find it a bit annoying that I have to specify a type for enums since often it really doesn't matter.

Reply:
umm - I'm talking about flatbuffer schema types - is this about C++ enums being typed (I have given up keeping track of that language) ?
In C the enums are represented as typecast integer defines in order to associate a type, .e.g. `#define Red INT32_C(1)`
Originally the were constants like `const int32_t Red = 1;` but that did not work in switch statements.

Reply:
I didn't like the `printf` in there either, therfore I moved the check for scoped-enums into the parser along with a check if it generates code for C++.

@mikkelfj C++11 introduced the following syntax `enum A : uint8_t { }`, i.e. `sizeof(A) == sizeof(uint8_t)`, unfortunately old compilers do not support it, hence the requirement for `--scoped-enums`

Reply:
OK, thanks

Reply:
The now removed third parameter in the `Array` template class was intended for this purpose, but it was indeed a bit messy as @vglavnyy pointed out.

Reply:
Just accidentally found this http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2008.pdf
It appers C might get typed enum in C2x, not that it matters since it would not be portable.

Reply:
@vglavnyy if it is needed to keep things simple, then so be it. I'd still like the actual error checking to be moved to the parser though, if possible.

Reply:
@aardappel I already moved it into the parser. I had to squash the history as rebasing was not fun otherwise.

Reply:
Ah thanks!